File: leetcode/python/1963.minimum-number-of-swaps-to-make-the-string-balanced.py
class Solution:
    def minSwaps(self, s: str) -> int:
        stack = []
        swaps = 0
        swapPosition = None

        for i, c in enumerate(s):
            if c == '[':
                stack.append(c)
            elif stack:
                stack.pop()
            else:
                if swapPosition is None:
                    swaps += 1
                    swapPosition = i
                else:
                    swapPosition = None
        
        return swaps

""" """

class Solution:
    def minSwaps(self, s: str) -> int:
        stack = 0
        swaps = 0
        swapPosition = None

        for i, c in enumerate(s):
            if c == '[':
                stack += 1
            elif stack:
                stack -= 1
            else:
                if swapPosition is None:
                    swaps += 1
                    swapPosition = i
                else:
                    swapPosition = None
        
        return swaps

File: leetcode/go/1963.minimum-number-of-swaps-to-make-the-string-balanced.go
func minSwaps(s string) int {
    open, swapPosition, swaps := 0, -1, 0
    for i, c := range s {
        if c == '[' {
            open++
        } else if open > 0 {
            open--
        } else {
            if swapPosition < 0 {
                swapPosition = i
                swaps++
            } else {
                swapPosition = -1
            }
        }
    }
    return swaps
}

